src/libfetchers/git.cc

```diff
@@ -406,49 +448,96 @@ struct GitInputScheme : InputScheme
+                        "Please make sure that the " ANSI_BOLD "rev" ANSI_NORMAL " exists on the "
+                        ANSI_BOLD "ref" ANSI_NORMAL " you've specified or add " ANSI_BOLD
+                        "allRefs = true;" ANSI_NORMAL " to " ANSI_BOLD "fetchGit" ANSI_NORMAL ".",
+                    input.getRev()->gitRev(),
+                    *input.getRef(),
+                    actualUrl
+                );
+            }
         }
 
         if (submodules) {
             Path tmpGitDir = createTempDir();
             AutoDelete delTmpGitDir(tmpGitDir, true);
 
+            // For this checkout approach, we need a commit, not just a treeish.
+            if (isDirtyTree) {
+                RunOptions gitOptions("git", { "-C", actualUrl, "commit-tree", "-m", "temporary commit for dirty tree", input.getRev()->gitRev() });
+                auto result = runProgram(gitOptions);
+                if (result.first)
+                    throw ExecError(result.first, fmt("program git commit-tree %1%", statusToString(result.first)));
+                input.attrs.insert_or_assign("rev", trim(result.second));
+            }
+
             runProgram("git", true, { "init", tmpDir, "--separate-git-dir", tmpGitDir });
             // TODO: repoDir might lack the ref (it only checks if rev
             // exists, see FIXME above) so use a big hammer and fetch
             // everything to ensure we get the rev.
             runProgram("git", true, { "-C", tmpDir, "fetch", "--quiet", "--force",
-                                      "--update-head-ok", "--", repoDir, "refs/*:refs/*" });
+                                      "--update-head-ok", "--", repoDir, "refs/*:refs/*",
+                                      input.getRev()->gitRev() });
 
             runProgram("git", true, { "-C", tmpDir, "checkout", "--quiet", input.getRev()->gitRev() });
             runProgram("git", true, { "-C", tmpDir, "remote", "add", "origin", actualUrl });
             runProgram("git", true, { "-C", tmpDir, "submodule", "--quiet", "update", "--init", "--recursive" });
 
+            if (dirtySubmoduleDiff.size()) {
+                RunOptions gitOptions("git", { "-C", tmpDir, "apply" });
+                StringSource s(dirtySubmoduleDiff);
+                gitOptions.standardIn = &s;
+                auto result = runProgram(gitOptions);
+                if (result.first)
+                    throw ExecError(result.first, fmt("program git apply %1%", statusToString(result.first)));
+            }
+
             filter = isNotDotGitDirectory;
         } else {
+            Strings noSmudgeOptions;
+            {
+                RunOptions gitOptions("git", { "-C", repoDir, "config", "-l" });
+                auto result = runProgram(gitOptions);
+                auto ss = std::stringstream{result.second};
+                StringSet filters;
+
+                for (std::string line; std::getline(ss, line, '\n');) {
```

Comment:
git config values can have multiple lines (try `git -c $'foo=two\nlines' config -l` to see for yourself), so splitting lines isn't sufficient here. I suggest to use `git config --name-only --get-regexp '^filter\..*\.smudge$'` to avoid this issue (config keys cannot contain newlines) and, as a bonus, only return relevant names.

Additionally, it isn't necessary to set the smudge filter to `cat`. If the filter is set to an empty string then no filtering is done.

With these two changes, `noSmudgeOptions` can be constructed by just appending `=` to the names returned from the `git config` invocation.